internal/appconfig/machines.go

```diff
@@ -62,6 +62,72 @@ func (c *Config) ToReleaseMachineConfig() (*fly.MachineConfig, error) {
 	return mConfig, nil
 }
 
+func (c *Config) ToTestMachineConfig(svc *ServiceMachineCheck, origMachine *fly.Machine) (*fly.MachineConfig, error) {
+	var machineEntrypoint []string
+	if len(svc.Entrypoint) > 0 {
+		machineEntrypoint = svc.Entrypoint
+	} else {
+		machineEntrypoint = origMachine.Config.Init.Entrypoint
+	}
+
+	var machineCommand []string
+	if len(svc.Command) > 0 {
+		machineCommand = svc.Command
+	} else {
+		return nil, fmt.Errorf("missing command for test machine")
+	}
+
+	var machineImage string
+	if svc.Image != "" {
+		machineImage = svc.Image
+	} else {
+		machineImage = origMachine.Config.Image
+	}
+
+	mConfig := &fly.MachineConfig{
+		Init: fly.MachineInit{
+			Cmd:        machineCommand,
+			SwapSizeMB: c.SwapSizeMB,
+			Entrypoint: machineEntrypoint,
+		},
+		Image: machineImage,
+		Restart: &fly.MachineRestart{
+			Policy: fly.MachineRestartPolicyNo,
+		},
+		AutoDestroy: true,
+		DNS: &fly.DNSConfig{
+			SkipRegistration: true,
+		},
+		Metadata: map[string]string{
+			fly.MachineConfigMetadataKeyFlyctlVersion:      buildinfo.Version().String(),
+			fly.MachineConfigMetadataKeyFlyPlatformVersion: fly.MachineFlyPlatformVersion2,
+			fly.MachineConfigMetadataKeyFlyProcessGroup:    fly.MachineProcessGroupFlyAppTestMachineCommand,
+		},
+		Env: lo.Assign(c.Env),
+	}
+
+	if c.Experimental != nil {
+		mConfig.Init.Entrypoint = c.Experimental.Entrypoint
+	}
+
+	mConfig.Env["FLY_TEST_COMMAND"] = "1"
+	mConfig.Env["FLY_PROCESS_GROUP"] = fly.MachineProcessGroupFlyAppTestMachineCommand
+	if c.PrimaryRegion != "" {
+		mConfig.Env["PRIMARY_REGION"] = c.PrimaryRegion
+	}
+	mConfig.Env["FLY_TEST_MACHINE_IP"] = lo.Ternary(origMachine == nil, "", origMachine.PrivateIP)
+
+	// Use the stop config from the app config by default
+	c.tomachineSetStopConfig(mConfig)
+
+	mConfig.StopConfig = &fly.StopConfig{
+		Timeout: lo.Ternary(svc.KillTimeout != nil, svc.KillTimeout, c.KillTimeout),
+		Signal:  lo.Ternary(svc.KillSignal != "", &svc.KillSignal, c.KillSignal),
```

Comment:
I think we want to default to top level KillSignal/Timeout values only when `svc.Image` isn't set.

Reply:
Hm, how come?

Reply:
let say the application needs a minute to drain connections and gracefully stop; also it stop the shutdown procedure on SIGUSR1 signal (not SIGTERM which is the default). 

Now the app also has a few machine checks, one uses the application image and others an external image like `curlimages/curl` 


```toml
app = "myapp"
kill_timeout = "1m"
kill_signal = "SIGUSR1"

[http_service]
internal_port = 8080

[[http_service.machine_checks]]
command = "/my-app runs-my-check"

[[http_service.machine_checks]]
image = "curlimages/curl"
command = ["-sf", "http://$FLY_TEST_MACHINE_IP:8080"]

```

Now, think about this:
1. Would you want the curl check to wait for a minute because that is what the app needs to gracefully drain and stop?
2. Is sending SIGUSR1 to curl's machine check what the user would expect us doing? I think this is a big gotcha moment.
3. Even for the first check that reuses the application image, does it makes sense to wait for a minute?
4. Do we need a way to replace FLY_TEST_MACHINE_IP by its actual value before launching the machine?
5. It is possible to rely on shell variable expansion but only for images that pipes the command trough a shell first

Reply:
that makes total sense, you've convinced me :)